### PR TITLE
expression, util: refactor type resolver using new design

### PR DESCRIPTION
**Pull request by XuHuaiyu**
refactor type resolver using the new design refer to [doc](https://docs.google.com/document/d/1YuftfsNJTmplhwNAnSgeYVk-ZCCyBrEOaCorL593yNI/edit#)

PTAL @hanfei1991 @shenli

### CODE REVIEW

File: expression/scalar_function.go [hunk 1]
```diff
@@ -146,7 +146,7 @@ func (sf *ScalarFunction) Decorrelate(schema *Schema) Expression {
 }
 
 // Eval implements Expression interface.
-func (sf *ScalarFunction) Eval(row []types.Datum) (types.Datum, error) {
```
Comment (hanfei1991) [inline expression/scalar_function.go]: Why change this

File: expression/expression.go [hunk 2]
```diff
@@ -172,10 +169,11 @@ func evalExprToString(expr Expression, row []types.Datum, _ *variable.StatementC
 	}
 	tc := expr.GetType().ToClass()
 	if tc == types.ClassString {
-		return val.GetString(), false, nil
+		// here we cannot use val.GetString() directly,
```
Comment (shenli) [inline expression/expression.go]: Why？

Comment (XuHuaiyu) [inline expression/expression.go]: @shenli 
For example, `Bit` will be regarded as ClassString in runtime refer to [MySQL](https://github.com/mysql/mysql-server/blob/5.7/sql/field.cc#L1069),
we store Bit in [Datum.i](https://github.com/pingcap/tidb/blob/master/util/types/datum.go#L218), while Datum.GetString() gets the value from [Datum.b](https://github.com/pingcap/tidb/blob/master/util/types/datum.go#L1344) directly.

Comment (shenli) [inline expression/expression.go]: OK, please add comment.

**Comment by XuHuaiyu** [comment]
PTAL @shenli

**Comment by shenli** [comment]
LGTM

**Pull request merged**